Neural style example

anmolsjoshi opened the pull request:
Pull Request for PyTorch Ignite implementation of Fast Neural Style. 

To run on Travis, the code creates a random style image, test image, content image and creates a model with CIFAR10 dataset. 

To correctly create a style transfer model, the MSCOCO dataset and an appropriate style image should be used. 

 For training
  - python examples/fast_neural_style/neural_style.py train --epochs 1 --cuda 0 --style_image random --test_image random
  
For evaluation
  - python examples/fast_neural_style/neural_style.py eval --content_image random --output_image /tmp/test.png --cuda 0 --model /tmp/checkpoints/checkpoint_net_1.pth

vfdev-5 commented:
@anmolsjoshi Thanks for the PR! 

I left some comments in the code on how to improve it. Next I'll check it on my machine etc and keep you updated.

Binary files :  examples/fast_neural_style/images/stylized_test/stylized_sample_epoch_000X.png are
produced by the example ? If yes we need to remove them too.

anmolsjoshi commented on .idea/workspace.xml:
```diff
@@ -0,0 +1,198 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Incorporated.

vfdev-5 commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,291 @@
+    def update_logs(engine):
+        for k, v in engine.state.output.items():
+            old_v = running_avgs.get(k, v)
+            new_v = alpha * old_v + (1 - alpha) * v
```
Can we use ignite metrics for that ?

anmolsjoshi replied:
Yes, we can. I got that from the dcgan example. I'll look into that this weekend.

anmolsjoshi replied:
@vfdev-5 Looking into this more, I'm not exactly sure how to incorporate Ignite Metrics into this. 

From what I have seen of the Metrics documentation, metrics are calculated using the output of the `step function` of the engine. In this case the metrics are the content_loss, style_loss and total_loss (MSE) that are needed for the optimization step inside the `step(engine, batch)` function. 

There are no metrics calculated at the end of Epoch, like in a classification problem.

This is why I used the `update_logs(engine)` function as shown. 

Happy to run with any ideas you may have.

vfdev-5 replied:
@anmolsjoshi I see your point, true that here there are running loss function values and we can not use Metrics in this case.

vfdev-5 commented on examples/fast_neural_style/README.md:
```diff
@@ -0,0 +1,76 @@
+
+It uses pytorch-ignite to implement an algorithm for artistic style transfer as described in [Perceptual Losses for Real-Time Style Transfer and Super-Resolution](https://arxiv.org/abs/1603.08155).
+
+This repo gives examples on how to create a Progress Bar with `pytorch-ignite`. 
```
What do you mean by : "how to create a Progress Bar with `pytorch-ignite`" ?

anmolsjoshi replied:
My bad, I can remove that. I added some functions in the code to create a progress bar to print metrics related to training. It was an additional example of how to use Events.

vfdev-5 commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,291 @@
+from torchvision import transforms
+
+from ignite.engine import Engine, Events
+from ignite.handlers import ModelCheckpoint, Timer
```
If you are not using Timer, it seems, we can remove it

anmolsjoshi replied:
Incorporated

vfdev-5 commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,291 @@
+import argparse
+import os
+import sys
+import time
```
Same here, it seems that time, re are not used, according to PyCharm

anmolsjoshi replied:
Incorporated

vfdev-5 commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,291 @@
+
+        bar = '[' + '=' * equal_to + '>' * progress + ' ' * (10 - equal_to - sub) + ']'
+
+        message = 'Epoch {epoch} | {percent_seen:.2f}% | {bar}'.format(epoch=engine.state.epoch,
```
Concerning your progress bar, looks cool on the execution. Could you please refactor it in a separate function ? Such that it could be reused by users if they would like to

anmolsjoshi replied:
Added handlers.py with Progbar class. The code is now a lot cleaner. And I was wondering if it could be added to the ignite handlers? Maybe that's a separate issue.

vfdev-5 replied:
@anmolsjoshi Look good ! Yes, I thought about to propose a sort of RunningMetrics handler. However in this case we would need to separate running metric computing and progress bar logic. For a progress bar reporting, maybe we could take a look at tqdm direction.

anmolsjoshi commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,291 @@
+
+    train_arg_parser = subparsers.add_parser("train", help="parser for training arguments")
+    train_arg_parser.add_argument("--epochs", type=int, default=2, help="number of training epochs, default is 2")
+    train_arg_parser.add_argument("--batch_size", type=int, default=8,
```
Incorporated

vfdev-5 commented:
Looks good to me!  Tested locally, looks ok

![test](https://user-images.githubusercontent.com/2459423/43684116-5e173722-989a-11e8-98aa-a9478e02dc69.png)

@anmolsjoshi nice solution for travis!

cc @alykhantejani @elanmart could you give your opinion

elanmart commented:
I haven't reviewed the algorithm, but the code and docs look good to me!

alykhantejani commented on examples/fast_neural_style/README.md:
```diff
@@ -0,0 +1,74 @@
+### Introduction
+This example is ported over from [pytorch-examples](https://github.com/pytorch/examples/tree/master/fast_neural_style).
+
+It uses pytorch-ignite to implement an algorithm for artistic style transfer as described in [Perceptual Losses for Real-Time Style Transfer and Super-Resolution](https://arxiv.org/abs/1603.08155).
```
nit: can we change `pytorch-ignite` to `ignite`. `pytorch-ignite` is the name on pip but the package is called `ignite`

anmolsjoshi replied:
Incorporated

alykhantejani commented on examples/fast_neural_style/README.md:
```diff
@@ -0,0 +1,74 @@
+
+### Requirements
+
+The code can be executed using both `Python 2.7` and `Python 3.5`.
```
this can be removed as is an assumption made by pytorch + ignite in general

anmolsjoshi replied:
Incorporated

alykhantejani commented on examples/fast_neural_style/README.md:
```diff
@@ -0,0 +1,74 @@
+
+* `torch`
+* `torchvision`
+* `pytorch-ignite`
```
should be `ignite` as it's the package name`

anmolsjoshi replied:
Incorporated

alykhantejani commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,249 @@
+    args = main_arg_parser.parse_args()
+
+    if args.subcommand is None:
+        print("ERROR: specify either train or eval")
```
these should be changed to `raise ValueError` instead of `print` and `sys.exit`

anmolsjoshi replied:
Incorporated

alykhantejani commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,249 @@
+
+    if args.content_image != 'test':
+        content_image = utils.load_image(args.content_image, scale=args.content_scale)
+    else:
```
we should remove this testing code from here. We don't actually need to do the smoke test and see results

anmolsjoshi replied:
Incorporated

alykhantejani commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,249 @@
+        transforms.Lambda(lambda x: x.mul(255))
+    ])
+
+    if args.dataset == 'test':
```
we should remove the testing code

anmolsjoshi replied:
Incorporated

alykhantejani commented:
Thanks for the PR.

In general this looks ok, I have some comments though:

1. We should remove the test code. I don't think its necessary to run a test on some fake data - we never look at the output images when this is run on travis.
2. I don't think we should do much work on travis either. I think the whole point of the smoke test is to run the code once (1 iteration max) just to make sure any API changes haven't broken the example. Anything longer than a few seconds is too long
3. We should remove the committed images, they are currently not displayed in the README and we should remove the test code anyway. Alternatively, we can display them in the README like the [original example](https://github.com/pytorch/examples/tree/master/fast_neural_style)
4. I'm not sure the ProgressBar is the best way to do this (we should be able to find a way to use tqdm), but as its for an example I'm ok with this

vfdev-5 commented:
@alykhantejani as discussed on slack, the time consumption on travis by this example is ~35 secs which is not the longest one among others.  So, we should find out a solution to all of the `if test` in the code as it serves only to run example on travis. Points 1) and 2) are to reconsider.

anmolsjoshi commented:
@vfdev-5 @alykhantejani I've incorporated all the comments. Quick question before I push another commit. Removing the testing code will have this fail on Travis because MSCOCO dataset or the Fake Dataset is needed to run the code, and downloading MSCOCO will take a while on Travis. 

If I do remove the test code, will it still have to pass Travis tests before completing the pull request? Or can it be incorporated without passing travis?

apologies if this is a novice question.

vfdev-5 commented:
@anmolsjoshi Sorry, I haven't yet time to writeup a guide on what we have decided on travis tests. 
Accepted solution is not to remove smoke tests and change the code such that to copy what is done in dcgan example with [`check_dataset`](https://github.com/pytorch/ignite/blob/master/examples/gan/dcgan.py#L156) to avoid urgly `if test` conditions in `train`. In check_dataset here we can similary support 'folder', 'mscoco', 'fake'.
As  `eval` does not use ignite we can avoid its testing.

vfdev-5 commented:
@anmolsjoshi in the README.md when I commited some of my modifications to this PR, I broke the correct representation of MSCOCO tree view. Sorry. Could you please restore it. Thanks

anmolsjoshi commented:
@vfdev-5 I fixed it! Here is a summary of changes:

- Updated README.md as requested. Added images as shown in pytorch examples.
- Removed function to stylize on epoch completion, I can add it back in, didn't think it was necessary. 
- Included check_dataset as in dcan.py example
- test works by creating Fake Dataset sized of one batch, thus test only uses one iteration
- Removed eval test from .travis.yml 
- Incorporated remaining comments as requested by @vfdev-5  and @alykhantejani

alykhantejani commented on examples/fast_neural_style/neural_style.py:
```diff
@@ -0,0 +1,216 @@
+
+        n_batch = len(x)
+
+        transformer.zero_grad()
```
`transformer.zero_grad()` and `optimizer.zero_grad()` should be the same AFAIK, but can we change this to `optimizer.zero_grad()` to be consistent with the other examples?

alykhantejani commented:
LGTM - @vfdev-5 if the changes are good with you lets go ahead and merge.

Thanks @anmolsjoshi!
